File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/ProjectConverterTest.java
package de.maibornwolff.codecharta.importer.scmlogparser;

import de.maibornwolff.codecharta.model.Project;
import de.maibornwolff.codecharta.model.input.Commit;
import de.maibornwolff.codecharta.model.input.VersionControlledFile;
import org.junit.Test;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Collections;

import static org.assertj.core.api.Assertions.assertThat;

public class ProjectConverterTest {

    @Test
    public void canCreateAnEmptyProject() throws Exception {
        // given
        String projectname = "Projectname";

        // when
        Project project = ProjectConverter.convert(projectname, Collections.emptyList(), true);

        //then
        assertThat(project.getNodes()).hasSize(0);
        assertThat(project.getProjectName()).isEqualTo(projectname);
    }

    @Test
    public void canConvertProjectWithAuthors() {
        //given
        String projectname = "ProjectWithAuthors";
        VersionControlledFile file1 = new VersionControlledFile("File 1");
        file1.registerCommit(new Commit("Author", Arrays.asList("File 1, File 2"), LocalDateTime.now()));

        //when
        Project project = ProjectConverter.convert(projectname, Arrays.asList(file1), true);

        //then
        assertThat(project.getRootNode().getChildren().get(0).getAttributes().containsKey("authors")).isTrue();
    }

    @Test
    public void canConvertProjectWithoutAuthors() {
        //given
        String projectname = "ProjectWithoutAuthors";
        VersionControlledFile file1 = new VersionControlledFile("File 1");
        file1.registerCommit(new Commit("Author", Arrays.asList("File 1, File 2"), LocalDateTime.now()));

        //when
        Project project = ProjectConverter.convert(projectname, Arrays.asList(file1), false);

        //then
        assertThat(project.getRootNode().getChildren().get(0).getAttributes().containsKey("authors")).isFalse();
    }


}

File: analysis/import/SCMLogParserV2/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparserv2/converter/ProjectConverterTest.kt
package de.maibornwolff.codecharta.importer.scmlogparserv2.converter

import de.maibornwolff.codecharta.importer.scmlogparserv2.input.Commit
import de.maibornwolff.codecharta.importer.scmlogparserv2.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparserv2.input.VersionControlledFile
import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.MetricsFactory
import de.maibornwolff.codecharta.importer.scmlogparserv2.parser.VersionControlledFilesList
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.time.OffsetDateTime

class ProjectConverterTest {

    private val metricsFactory = mockk<MetricsFactory>()

    @Before
    fun setup() {
        every { metricsFactory.createMetrics() } returns listOf()
    }

    private fun modificationsByFilename(vararg filenames: String): List<Modification> {
        return filenames.map { Modification(it) }
    }

    @Test
    @Throws(Exception::class)
    fun canCreateAnEmptyProject() {
        // given
        val projectConverter = ProjectConverter(true)
        val metricsFactory = MetricsFactory()
        val vcfList = VersionControlledFilesList(metricsFactory)
        // when
        val project = projectConverter.convert(vcfList, metricsFactory, listOf())

        // then
        assertThat(project.rootNode.leaves).hasSize(1)
    }

    @Test
    fun canConvertProjectWithAuthors() {
        // given
        val projectConverter = ProjectConverter(true)
        val file1 = VersionControlledFile("File 1", metricsFactory)
        val vcfList = VersionControlledFilesList(metricsFactory)

        vcfList.addFileBy("File 1")
        file1.registerCommit(
            Commit("Author", modificationsByFilename("File 1", "File 2"), OffsetDateTime.now()),
            Modification.EMPTY
        )

        // when
        val project = projectConverter.convert(vcfList, metricsFactory, listOf("File 1"))

        // then
        assertThat(project.rootNode.children.toMutableList()[0].attributes.containsKey("authors")).isTrue()
    }

    @Test
    fun canConvertProjectWithoutAuthors() {
        // given
        val projectConverter = ProjectConverter(false)
        val file1 = VersionControlledFile("File 1", metricsFactory)
        val vcfList = VersionControlledFilesList(metricsFactory)

        vcfList.addFileBy("File 1")
        file1.registerCommit(
            Commit("Author", modificationsByFilename("File 1", "File 2"), OffsetDateTime.now()),
            Modification.EMPTY
        )

        // when
        val project = projectConverter.convert(vcfList, metricsFactory, listOf("File 1"))

        // then
        assertThat(project.rootNode.children.toMutableList()[0].attributes.containsKey("authors")).isFalse()
    }

    @Test
    fun attributeTypesAreCreated() {
        val projectConverter = ProjectConverter(false)
        val vcfList = VersionControlledFilesList(metricsFactory)
        val project = projectConverter.convert(vcfList, MetricsFactory(), listOf())

        assertThat(project.attributeTypes).containsKeys("edges", "nodes")
    }
}
